controllers/snapshotenvironmentbinding_controller.go

```diff
@@ -0,0 +1,194 @@
+
+	var snapshotEnvBinding appstudiov1alpha1.SnapshotEnvironmentBinding
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, &snapshotEnvBinding); err != nil {
+		if errors.IsNotFound(err) {
+			lg.V(logs.DebugLevel).Info("SnapshotEnvironmentBinding already gone from the cluster. skipping reconciliation")
+			return ctrl.Result{}, nil
+		}
+
+		return ctrl.Result{}, fmt.Errorf("failed to get the SnapshotEnvironmentBinding: %w", err)
+	}
+
+	finalizationResult, err := r.finalizers.Finalize(ctx, &snapshotEnvBinding)
+	if err != nil {
+		// if the finalization fails, the finalizer stays in place, and so we don't want any repeated attempts until
+		// we get another reconciliation due to cluster state change
+		return ctrl.Result{Requeue: false}, fmt.Errorf("failed to finalize: %w", err)
+	}
+	if finalizationResult.Updated {
+		if err = r.k8sClient.Update(ctx, &snapshotEnvBinding); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update snapshot env binding based on finalization result: %w", err)
+		}
+	}
+	if finalizationResult.StatusUpdated {
+		if err = r.k8sClient.Status().Update(ctx, &snapshotEnvBinding); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update the snapshot env binding status based on finalization result: %w", err)
+		}
+	}
+
+	if snapshotEnvBinding.DeletionTimestamp != nil {
+		lg.V(logs.DebugLevel).Info("SnapshotEnvironmentBinding is being deleted. skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
+
+	appNamespace := snapshotEnvBinding.Namespace
+	environmentName := snapshotEnvBinding.Spec.Environment
+	applicationName := snapshotEnvBinding.Spec.Application
+
+	// Get the Environment CR
+	environment := appstudiov1alpha1.Environment{}
+	err = r.k8sClient.Get(ctx, types.NamespacedName{Name: environmentName, Namespace: snapshotEnvBinding.Namespace}, &environment)
+	if err != nil {
+		lg.Error(err, fmt.Sprintf("unable to get the Environment %s %v", environmentName, req.NamespacedName))
+		return ctrl.Result{}, err
+	}
+
+	target, err := detectTargetFromEnvironment(ctx, environment)
+
+	remoteSecretsList := api.RemoteSecretList{}
+	if err := r.k8sClient.List(ctx, &remoteSecretsList, client.InNamespace(appNamespace), client.MatchingLabels{ApplicationLabelName: applicationName, EnvironmentLabelName: environmentName}); err != nil {
+		spiFileContentRequestLog.Error(err, "Unable to fetch remote secrets list", "namespace", appNamespace)
+		return ctrl.Result{}, unableToFetchRemoteSecretsError
+	}
+
+	for rs := range remoteSecretsList.Items {
+		remoteSecret := remoteSecretsList.Items[rs]
+		addTargetIfNotExists(&remoteSecret, target)
+		if err = r.k8sClient.Update(ctx, &remoteSecret); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to update secret: %w", err)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func addTargetIfNotExists(secret *api.RemoteSecret, target api.RemoteSecretTarget) {
+	for idx := range secret.Spec.Targets {
+		existingTarget := secret.Spec.Targets[idx]
```

Comment:
I think this kind of check is excessive, it is RemoteSecret's business to care about its integrity.

Reply:
The controllers should not update the spec, because the spec should express the user's intent. So the RS controller would just leave the duplicate entries in the spec and just document in the status that they are duplicates and that it is ignoring them.

In another words, I would keep this sanity check in place.